presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java

```diff
@@ -244,6 +244,51 @@ public void testHashAggregationWithGlobals(boolean hashEnabled, long memoryLimit
         assertOperatorEqualsIgnoreOrder(operatorFactory, driverContext, input, expected, hashEnabled, Optional.of(groupByChannels.size()));
     }
 
+    @Test(dataProvider = "hashEnabledAndMemoryLimitBeforeSpillValues")
+    public void testHashAggregationMemoryReservation(boolean hashEnabled, long memoryLimitBeforeSpill, long memoryLimitForMergeWithMemory)
+            throws Exception
+    {
+        MetadataManager metadata = MetadataManager.createTestMetadataManager();
+        InternalAggregationFunction arrayAggColumn = metadata.getFunctionRegistry().getAggregateFunctionImplementation(
+                new Signature("array_agg", AGGREGATE, parseTypeSignature("array(bigint)"), parseTypeSignature(StandardTypes.BIGINT)));
+
+        List<Integer> hashChannels = Ints.asList(1);
+        RowPagesBuilder rowPagesBuilder = rowPagesBuilder(hashEnabled, hashChannels, BIGINT, BIGINT);
+        List<Page> input = rowPagesBuilder
+                .addSequencePage(10, 100, 0)
+                .addSequencePage(10, 200, 0)
+                .addSequencePage(10, 300, 0)
+                .build();
+
+        DriverContext driverContext = createTaskContext(executor, TEST_SESSION, new DataSize(10, Unit.MEGABYTE))
+                .addPipelineContext(0, true, true)
+                .addDriverContext();
+
+
+        HashAggregationOperatorFactory operatorFactory = new HashAggregationOperatorFactory(
+                0,
+                new PlanNodeId("test"),
+                ImmutableList.of(BIGINT),
+                hashChannels,
+                ImmutableList.of(),
+                Step.SINGLE,
+                true,
+                ImmutableList.of(arrayAggColumn.bind(ImmutableList.of(0), Optional.empty())),
+                rowPagesBuilder.getHashChannel(),
+                Optional.empty(),
+                100_000,
+                new DataSize(16, MEGABYTE),
+                memoryLimitBeforeSpill > 0,
```

Comment:
please make sure to have `memoryLimitBeforeSpill` values that will both enable and disable spill, so we test both in memory and spillable hash aggregation builders.